### PR TITLE
Add RecordFieldExtractor feature

### PR DESCRIPTION
**Pull request by Toilal**
This new feature allows more control on some RecordMarshaller implementation (Flat file, CSV, ...) to build the list of field from the record.

### CODE REVIEW

File: easybatch-core/src/main/java/org/easybatch/core/api/RecordFieldExtractionException.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ *   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *   THE SOFTWARE.
+ */
+
+package org.easybatch.core.api;
+
+/**
+ * Exception thrown by {@link RecordFieldExtractor} implementations.
+ *
+ * @author Mahmoud Ben Hassine (mahmoud@benhassine.fr)
```
Comment (benas) [inline easybatch-core/src/main/java/org/easybatch/core/api/RecordFieldExtractionException.java]: Do not hesitate to put your name as author of new APIs you introducde

Comment (Toilal) [inline easybatch-core/src/main/java/org/easybatch/core/api/RecordFieldExtractionException.java]: No problem :)

**Comment by benas** [comment]
I'm not sure I understand well the purpose of the `RecordFieldExtractor` API. Is it for having more control over the order in which properties of the record will be written? May be a unit test of the `BeanRecordFieldExtractor` will help me out

**Comment by Toilal** [comment]
Yes, to have more control on ordering, and to support an object than more complex than a simple flat JavaBean.

Here is an example of a custom implementation :

```java
public class InteractionFieldExtractor implements RecordFieldExtractor<InteractionEntity> {
    private Object getEntity(ReferenceEntity entity) {
        return entity != null ? entity.getId() : null;
    }

    @Override
    public Object[] extract(InteractionEntity item) {
        List<Object> list = new ArrayList<>();

        list.add(item.getId());
        list.add(item.getKeygen());
        list.add(item.getPersonalId());
        list.add(item.getContractId());
        list.add(item.getUnifiedPartyId());
        list.add(item.getCreationTimestamp());
        list.add(item.getCreationDate());
        list.add(item.getApplicationSourceId());
        list.add(item.getExternalSystemId());
        list.add(item.getCallNumber());
        list.add(item.getContactValue());
        list.add(item.getLastName());
        list.add(item.getFirstName());
        list.add(item.getOperatorId());
        list.add(item.getOperatorSkill());
        list.add(item.getOperatorTeamHierarchyId());
        list.add(item.getOperatorTeamActivityId());

        list.add(getEntity(item.getStatus()));
        list.add(item.getUniverse().getId());
        list.add(item.getDirection().getId());
        list.add(item.getChannel().getId());

        list.add(getEntity(item.getMedia()));
        list.add(item.getOrigin());

        list.add(item.getWay().getId());
        list.add(item.getReason().getId());
        list.add(item.getDirection().getId());
        list.add(getEntity(item.getReasonDetail()));


        list.add(item.getComment());
        list.add(item.getConclusion().getId());
        list.add(item.getLegalPosition().getId());
        list.add(getEntity(item.getPriority()));
        list.add(getEntity(item.getActivity()));
        list.add(getEntity(item.getRelatedObject()));
        list.add(item.getRelatedObjectId());
        list.add(item.getGedId());
        list.add(item.getProduct());
        list.add(item.getProductCategory());
        list.add(item.getEnttGestnId());
        list.add(item.getPhoneCallId());
        list.add(item.getLinkedId());
        list.add(item.getLinkedKeygen());
        list.add(item.getApplicationSourceId2());
        list.add(item.getApplicationSourceId2());

        return list.toArray();
    }

}
```

**Comment by Toilal** [comment]
The ```BeanRecordFieldExtractor``` is the default implementation that match the behavior of the current codebase. Only difference is when no field is given in the list, then it will automatically use all the properties of the bean.

**Comment by benas** [comment]
Ok, I see. Another advantage of this API is that the user can format the data that will be written.

That's fine, please update your name in the javadoc and add a unit test for the `BeanRecordFieldExtractor`. I'll update the documentation to explain the usage of this new API.

Great addition! Many thanks

**Comment by benas** [comment]
Is this API the same thing as spring batch's [FieldExtractor](http://docs.spring.io/spring-batch/trunk/apidocs/org/springframework/batch/item/file/transform/FieldExtractor.html) ?
Seems like the `BeanRecordFieldExtractor` is also the equivalent of spring batch's [BeanWrapperFieldExtractor](http://docs.spring.io/spring-batch/trunk/apidocs/org/springframework/batch/item/file/transform/BeanWrapperFieldExtractor.html)

**Comment by Toilal** [comment]
Yes, exactly the same idea.
Le 6 août 2015 17:23, "Mahmoud Ben Hassine" <notifications@github.com> a
écrit :

File: easybatch-core/src/main/java/org/easybatch/core/field/BeanRecordFieldExtractor.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import org.easybatch.core.api.RecordFieldExtractor;
+import org.easybatch.core.util.Utils;
+
+import java.beans.IntrospectionException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Map;
+
+/**
+ * Use JavaBean convention with Introspector to extractor properties values from records of a specified class.
```
Comment (benas) [inline easybatch-core/src/main/java/org/easybatch/core/field/BeanRecordFieldExtractor.java]: typo here, I guess you mean "to extract"

File: easybatch-core/src/test/java/org/easybatch/core/field/BeanRecordFieldExtractorTest.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+/**
+ * Test class for {@link BeanRecordFieldExtractor}.
+ *
+ * @author Rémi Alvergnat (toilal.dev@gmail.com)
+ */
+public class BeanRecordFieldExtractorTest {
```
Comment (benas) [inline easybatch-core/src/test/java/org/easybatch/core/field/BeanRecordFieldExtractorTest.java]: Perfect! Thank you for the update

**Comment by benas** [comment]
Perfect! Straight to the master branch :-)

**Pull request merged**